### ISSUE TITLE
Pytorch基本知识

### ISSUE THREAD

**Issue by Zener0710** [Zener0710/blogs#5]
# 初始化张量
默认情况下，张量在cpu上创建。借助 **tensor.to()** 函数，可以将张量转移至gpu存储。
1. 使用 **torch.tensor()** 函数，将原数据直接转化为张量
2. 使用 **torch.from_numpy()** 函数，将array格式的数据转化为张量
3. 使用 **torch.ones_like()** 等函数，按照已有张量A生成新的张量B。一般情况下，新张量会保留原张量的property（shape、datatype），除非在函数中指定新的相关参数
4. 使用 **torch.rand()** 等函数，在指定shape的情况下，随机生成新的张量

# 张量的属性
张量的属性（attribute）包括：shape、datatype、device on which they are stored

# 张量的运算
Pytorch的张量支持超过100种运算，包括数学运算、线性代数、matrix manipulation（transpose、index、slice）
每一种运算都既可以在cpu上执行，也可以在gpu上执行，取决于张量存储在哪个设备上

## 1. Index And Slice
>tensor = torch.ones(4, 4)
print('First row: ',tensor[0])
print('First column: ', tensor[:, 0])
print('Last column:', tensor[..., -1])
tensor[:,1] = 0

## 2. Concatenate
>torch.cat([tensor, tensor, tensor], dim=1)

## 3. Matrix Multiply
>y1 = tensor @ tensor.T
y2 = tensor.matmul(tensor.T)
y3 = torch.rand_like(tensor)
torch.matmul(tensor, tensor.T, out=y3)

## 4. Element-wise Multiply
>z1 = tensor * tensor
z2 = tensor.mul(tensor)
z3 = torch.rand_like(tensor)
torch.mul(tensor, tensor, out=z3)

## 5. Extract Python Numeric Value From _Single-element Tensor_
If you have a _**one-element tensor**_, you can convert it to a **_Python numerical value_** using item():
>agg = tensor.sum()
agg_item = agg.item()

## 6. In-place（原地）Operation
Operations that store the result into the operand are called in-place. They are denoted by a _ suffix. 
For example: x.copy_(y), x.t_(), will change x.
>tensor([[1., 0., 1., 1.],
        [1., 0., 1., 1.],
        [1., 0., 1., 1.],
        [1., 0., 1., 1.]])
tensor.add_(5)
tensor([[6., 5., 6., 6.],
        [6., 5., 6., 6.],
        [6., 5., 6., 6.],
        [6., 5., 6., 6.]])
